executor/builder.go

```diff
@@ -771,6 +771,16 @@ func (b *executorBuilder) buildUnionScanFromReader(reader Executor, v *plannerco
 		us.conditions = v.Conditions
 		us.columns = x.columns
 		err = us.buildAndSortAddedRows()
+	case *UnionExec:
+		// For partitioned table, table reader may be transformed to UnionAll.
+		if x.fromPartition != nil {
+			us.dirty = GetDirtyDB(b.ctx).GetDirtyTable(x.fromPartition.Meta().ID)
+			us.conditions = v.Conditions
+			us.columns = x.fromPartition.Meta().Cols()
```

Comment:
What about
- Check whether the child operator of `UnionScan` is `UnionAll`.
- If so, use its `Children[0]` instead.
- If not, just use itself.
- Use the `switch` clause of line 703.